### ISSUE TITLE
Failed to build from source on Debian armhf

### ISSUE THREAD

**Issue by rtsisyk** [docker/docker#20731]
scaleway.com host
Debian stretch/sid **armhf**
Linux xx 4.1.6-249 #1 SMP Mon Aug 31 12:34:47 UTC 2015 armv7l GNU/Linux
gcc version 5.3.1 20160220 (Debian 5.3.1-9)
go version go1.5.3 linux/arm

Debian has `docker.io` package 1.8.3~ds1-2:

```
docker version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   f4bf5c7
 Built:        
 OS/Arch:      linux/arm

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   f4bf5c7
 Built:        
 OS/Arch:      linux/arm
```

```
docker info
Containers: 10
Images: 70
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.6-249
Operating System: Debian GNU/Linux stretch/sid (containerized)
CPUs: 4
Total Memory: 1.975 GiB
Name: 
ID: EQRH:RQQ7:CKZN:KXGB:TML2:6462:C5X2:7FOT:OMTC:HTTT:DKMG:BJCC
Username: rtsisyk
Registry: https://index.docker.io/v1/
```

I want to upgrade to the latest version (1.10 or newer), because Debian package has troubles when work with the Docker Hub registry. I was unable to find armhf packages, so I tried to build them on locally.

First of all,  I tried ```docker build -t docker -f Dockerfile.armhf .```

The build failed on RUN ./contrib/download-frozen-image-v2.sh:

```
Step 31 : RUN ln -sv $PWD/contrib/completion/bash/docker /etc/bash_completion.d/docker
 ---> Using cache
 ---> 5bf03fd6aafd
Step 32 : COPY contrib/download-frozen-image-v2.sh /go/src/github.com/docker/docker/contrib/
 ---> Using cache
 ---> 170fee5e0ab3
Step 33 : RUN ./contrib/download-frozen-image-v2.sh /docker-frozen-images       armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb     armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4    armhf/debian:jessie@sha256:4a2187483f04a84f9830910fe3581d69b3c985cc045d9f01d8e2f3795b28107b     armhf/hello-world:latest@sha256:161dcecea0225975b2ad5f768058212c1e0d39e8211098666ffa1ac74cfb7791
 ---> Running in 4f11a2035e92
error: /v2/armhf/buildpack-deps/manifests/sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb returned something unexpected:
  <html><body><h1>408 Request Time-out</h1>
Your browser didn't send a complete request in time.
</body></html>
The command '/bin/sh -c ./contrib/download-frozen-image-v2.sh /docker-frozen-images     armhf/buildpack-deps:jessie@sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb     armhf/busybox:latest@sha256:d98a7343ac750ffe387e3d514f8521ba69846c216778919b01414b8617cfb3d4    armhf/debian:jessie@sha256:4a2187483f04a84f9830910fe3581d69b3c985cc045d9f01d8e2f3795b28107b     armhf/hello-world:latest@sha256:161dcecea0225975b2ad5f768058212c1e0d39e8211098666ffa1ac74cfb7791' returned a non-zero code: 1
```

IMHO, it was worse idea to require Docker  to build Docker.

Next, I installed Go and all dependencies on my host  using `apt-get build-dep docker.io` and tried to build Docker without Docker:

```
myhost# AUTO_GOPATH=1 ./hack/make.sh 

[CUT] 

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev

 # _/home/roman/docker/docker
/usr/lib/go/pkg/tool/linux_arm/link: running gcc failed: exit status 1
/usr/bin/ld: cannot find -ldevmapper
/usr/bin/ld: cannot find -ldevmapper
collect2: error: ld returned 1 exit status
```

```
lrwxrwxrwx 1 root root 56 Feb 20 13:43 /usr/lib/arm-linux-gnueabihf/libdevmapper-event-lvm2.so -> /lib/arm-linux-gnueabihf/libdevmapper-event-lvm2.so.2.02
lrwxrwxrwx 1 root root 53 Feb 20 13:43 /usr/lib/arm-linux-gnueabihf/libdevmapper-event.so -> /lib/arm-linux-gnueabihf/libdevmapper-event.so.1.02.1
lrwxrwxrwx 1 root root 47 Feb 20 13:43 /usr/lib/arm-linux-gnueabihf/libdevmapper.so -> /lib/arm-linux-gnueabihf/libdevmapper.so.1.02.1
```

```
dpkg --get-selections|grep devmapper
libdevmapper-dev:armhf                          install
libdevmapper-event1.02.1:armhf                  install
libdevmapper1.02.1:armhf                        install
```

How I can install Docker on my Debian?

**Comment by cpuguy83** [comment]
@rtsisyk It is not required to have docker to build docker, it is just a hell of a lot simpler since there are many many dependencies that need to be linked in.


The error here is pretty straight forward:
```
error: /v2/armhf/buildpack-deps/manifests/sha256:ca6cce8e5bf5c952129889b5cc15cd6aa8d995d77e55e3749bbaadae50e476cb returned something unexpected:
  <html><body><h1>408 Request Time-out</h1>
Your browser didn't send a complete request in time.
```

Why it's timing out I don't know.

In terms of just being able to build docker, you should be able to comment out this line. This bit should only be needed for the integration tests.

**Comment by rtsisyk** [comment]
> The error here is pretty straight forward:

Obviously, but I can't do anything with it. My network connection is strong and I tried multiple times.
It seems Docker 1.8.3 doesn't work well with the docker registry. This is a reason I want to upgrade.

>In terms of just being able to build docker, you should be able to comment out this line. This bit should only be needed for the integration tests.

works for me, thanks!

How it is possible to build `docker-engine` Debian package similar to packages from https://apt.dockerproject.org/repo? I tried `docker run --privileged docker hack/make.sh binary ubuntu`, but this script ended with `fpm: command not found`.
